enaml/core/dynamicscope.py

```diff
@@ -0,0 +1,107 @@
+# ------------------------------------------------------------------------------
+# Copyright (c) 2025, Nucleic Development Team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ------------------------------------------------------------------------------
+from functools import partial
+from itertools import chain, repeat
+from enaml.core.declarative import Declarative
+from enaml.core._dynamicscope import _DynamicScope, UserKeyError
+
+
+def d_iter(owner: Declarative):
+    """Iterate attribute names of the declarative and it's ancestors.
+
+    Parameters
+    ----------
+    owner: Declarative
+        The declarative to walk.
+
+    Yields
+    ------
+    name: string
+        The attribute name
+
+    """
+    while owner is not None:
+        for name in dir(owner):
+            yield name
+        owner = owner._parent
+
+
+def include_key(key: str, used: set) -> bool:
+    """Filter function to determine whether the key should be included in the
+    dynamicscope's iter results.
+
+    Parameters
+    ----------
+    key: string
+        The scope key.
+    used: set[str]
+        The set of keys already seen.
+
+    Returns
+    -------
+    result: bool
+        Whether the key should be included.
+    """
+    if key.startswith("__") or key in used:
```

Comment:
Should filter all keys with a single leading `_` instead of just double? i.e. only yield the "public" keys.